File: DesktopBrowser/Server/Utils/PathInfo.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Web;
using System.IO;

namespace DesktopBrowser.Server.Utils
{
    public class PathInfo
    {
        public string Value { get; private set; }

        public PathInfo(string path)
        {
            Value = path;
        }

        public bool IsRoot
        {
            get
            {
                if (IsEmpty)
                    return false;
                var abs = ToAbsolute();
                return System.IO.Path.GetPathRoot(abs.Value) == abs.Value;
            }
        }
        public bool IsEmpty
        {
            get
            {
                return Value.IsNullOrEmpty();
            }
        }
        public PathInfo Combine(string path)
        {
            if (IsEmpty)
                return new PathInfo(path);
            return new PathInfo(System.IO.Path.Combine(this.Value, path));
        }
        public PathInfo Combine(PathInfo path)
        {
            return Combine(path.Value);
        }

        public PathInfo ToAbsolute()
        {
            if (IsEmpty)
                return this;
            if (System.IO.Path.IsPathRooted(Value))
                return this;
            return new PathInfo(System.IO.Path.GetFullPath(Value));
        }
        public PathInfo ToAbsoluteExact()
        {
            if (IsEmpty)
                return this;
            return new PathInfo(new DirectoryInfo(Value).FullName);
        }

        public override string ToString()
        {
            return Value;
        }

        public bool IsFile
        {
            get { return System.IO.File.Exists(Value); }
        }
        public bool IsDirectory
        {
            get { return System.IO.Directory.Exists(Value); }
        }
        public PathInfo ParentPath
        {
            get
            {
                if (IsEmpty)
                    return this;
                if (IsRoot)
                    return new PathInfo("");
                return new PathInfo(System.IO.Path.GetDirectoryName(Value));
            }
        }

        public string Name
        {
            get
            {
                if (IsRoot)
                    return Value;
                return System.IO.Path.GetFileName(Value);
            }
        }


        public string Extension
        {
            get { return System.IO.Path.GetExtension(Value); }
        }
    }
}

File: server/src/utils/path-info.ts
import {IoFile, IoDir, IoPath, FileSystemInfo } from "./io"

export class PathInfo {
    Value: string;

    constructor(path: string) {
        this.Value = path;
    }

    get IsRoot(): boolean {
        if (this.IsEmpty)
            return false;
        var abs = this.ToAbsolute();
        return IoPath.GetPathRoot(abs.Value) == abs.Value;
    }
    get IsEmpty(): boolean {
        return String.isNullOrEmpty(this.Value);
    }
    Combine(path: PathInfo | string): PathInfo {
        if (typeof (path) == "string") {
            if (this.IsEmpty)
                return new PathInfo(path);
            return new PathInfo(IoPath.Combine(this.Value, path));
        }
        return this.Combine(path.Value);
    }

    ToAbsolute(): PathInfo {
        if (this.IsEmpty)
            return this;
        if (IoPath.IsPathRooted(this.Value))
            return this;
        return new PathInfo(IoPath.GetFullPath(this.Value));
    }
    async ToAbsoluteExact(): Promise<PathInfo> {
        if (this.IsEmpty)
            return this;
        return new PathInfo((await FileSystemInfo.create(this.Value)).FullName);
    }

    toString(): string {
        return this.Value;
    }

    get IsFile(): Promise<boolean> {
        return IoFile.Exists(this.Value);
    }
    get IsDirectory(): Promise<boolean> {
        return IoDir.Exists(this.Value);
    }
    get ParentPath(): PathInfo {
        if (this.IsEmpty)
            return this;
        if (this.IsRoot)
            return new PathInfo("");
        var x = this.Value;
        if (x.endsWith("\\"))
            x = x.removeLast(1);
        return new PathInfo(IoPath.GetDirectoryName(x));
    }

    get Name(): string {
        if (this.IsRoot)
            return this.Value;
        return IoPath.GetFileName(this.Value);
    }


    get Extension(): string {
        return IoPath.GetExtension(this.Value);
    }
}
